web-ui/src/app/components/tree-visualizer/tree-visualizer.js

```diff
@@ -446,41 +447,60 @@ var jQuery = require('jquery');
+                            right: $window.width() - targetRect.right,
+                            bottom: $window.innerHeight() - targetRect.bottom,
                         },
                         treeV = {
-                            right: $window.outerWidth() - treeRect.right,
-                            bottom: $window.outerHeight() - treeRect.bottom,
+                            right: $window.width() - treeRect.right,
+                            bottom: $window.innerHeight() - treeRect.bottom,
                         };
-
+                    // The maximum to the right the tooltip can be positioned without overlapping the viewport
+                    var rightestLeft = parseInt($window.width() - tooltip.outerWidth() - 20, 10)
                     var tooltipV = {
-                        left: parseInt(targetRect.left + (targetRect.width / 2) - (tooltip.outerWidth() / 2) + 8, 10) + "px",
+                        left: Math.max(10, Math.min(rightestLeft, parseInt(targetRect.left + (targetRect.width / 2) - (tooltip.outerWidth() / 2) + 8, 10))) + "px",
                         top: parseInt(targetRect.top - tooltip.outerHeight() - 24, 10)
                     }
+                    // Speech bubble arrow positioned independently from tooltip rectangle
+                    var arrowV = {
+                        left: parseInt(targetRect.left + (targetRect.width / 2) - (arrowRect.width / 2), 10), //the centre of the target node
+                        top: parseInt(tooltipV.top + tooltip.outerHeight() - 1) //the bottom of the tooltip balloon
+                    }
 
                     if (animate) {
                         tooltip.stop(true).animate({
                             "left": tooltipV.left,
                             "top": tooltipV.top
-                        }, 250);
+                        }, { duration: 250, queue: false });
+                        arrow.stop(true).animate({
+                            "left": arrowV.left,
+                            "top": arrowV.top
+                        }, { duration: 250, queue: false });
                     } else {
                         tooltip.css({
                             "left": tooltipV.left,
                             "top": tooltipV.top
                         });
+                        arrow.css({
+                            "left": arrowV.left,
+                            "top": arrowV.top
+                        });
                     }
 
                     if (((targetRect.left + (targetRect.width / 2)) < treeRect.left) ||
                         ((targetV.right + (targetRect.width / 2)) < treeV.right) ||
                         ((targetRect.top + (targetRect.height / 2)) < treeRect.top) ||
                         ((targetV.bottom + targetRect.height) < treeV.bottom)) {
+                        // the node is scrolled outside of the view
                         tooltip.fadeOut(400);
+                        arrow.fadeOut(400);
```

Comment:
Would it possible to place the arrow inside the tooltip? When the user closes the tooltip now, the arrow persists.

Reply:
![image](https://user-images.githubusercontent.com/24268578/45888200-c6686880-bdbd-11e8-99ff-10cc49098380.png)